src/task_processor/threads.py

```diff
@@ -95,18 +96,46 @@ def run(self) -> None:
             time.sleep(self.sleep_interval_millis / 1000)
 
     def run_iteration(self) -> None:
-        try:
-            run_tasks(self.queue_pop_size)
-            run_recurring_tasks()
-        except Exception as e:
-            # To prevent task threads from dying if they get an error retrieving the tasks from the
-            # database this will allow the thread to continue trying to retrieve tasks if it can
-            # successfully re-establish a connection to the database.
-            # TODO: is this also what is causing tasks to get stuck as locked? Can we unlock
-            #  tasks here?
-
-            logger.error("Received error retrieving tasks: %s.", e, exc_info=e)
-            close_old_connections()
+        """
+        Consume and execute tasks from the queue, and run recurring tasks
+
+        This method tries to consume tasks from multiple databases as to ensure
+        that any remaining tasks are processed after opting in or out of a
+        separate database setup.
+        """
+        database_is_separate = "task_processor" in settings.TASK_PROCESSOR_DATABASES
+        if database_is_separate:
+            assert (
+                "task_processor.routers.TaskProcessorRouter"
+                in settings.DATABASE_ROUTERS
+            ), (
+                "DATABASE_ROUTERS must include 'task_processor.routers.TaskProcessorRouter' "
+                "when using a separate task processor database."
+            )  # This is for our own sanity
+            assert "task_processor" in settings.DATABASES, (
+                "DATABASES must include 'task_processor' when using a separate task processor database."
+            )  # ¯\_(ツ)_/¯ One has to read the documentation and fix it: https://docs.flagsmith.com/deployment/configuration/task-processor
+
+        for database in settings.TASK_PROCESSOR_DATABASES:
+            try:
+                run_tasks(database, self.queue_pop_size)
+
+                # Recurring tasks are only run on one database
+                if (database == "default") ^ database_is_separate:
+                    run_recurring_tasks(database)
+            except Exception as exception:
+                # To prevent task threads from dying if they get an error retrieving the tasks from the
+                # database this will allow the thread to continue trying to retrieve tasks if it can
+                # successfully re-establish a connection to the database.
+                # TODO: is this also what is causing tasks to get stuck as locked? Can we unlock
```

Comment:
Unrelated: we can get rid of this todo now

Reply:
I think this is a bug and it still persists. 🤔  The [database function](https://github.com/Flagsmith/flagsmith-common/blob/9e224f121ffbb32d332f648d31b9089cacfce6c7/src/task_processor/migrations/sql/0011_get_tasks_to_process.sql) will lock all matching tasks regardless of they being consumed by the queryset. We might need a refactor of querying tasks if this is a pressing issue.

Reply:
Sorry, I should have provided more context. The comment was added when we were trying to figure out a different issue. This has never caused tasks to get stuck in the locked stage. If you’d prefer, feel free to create an issue highlighting this.